### PR TITLE
[New library] System UI Controller

### PR DESCRIPTION
**Pull request by chrisbanes**
This takes the work which @nickbutcher created in [Jetsnack](https://github.com/android/compose-samples/blob/main/Jetsnack/app/src/main/java/com/example/jetsnack/ui/utils/SystemUi.kt), and wraps it up in a library.

Used the new utility to tidy up the inset samples.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,6 +10,9 @@ A number of libraries which integrate popular image loading libraries into Jetpa
 ### 📐 [Insets](./insets/)
 A library which brings [WindowInsets](https://developer.android.com/reference/kotlin/android/view/WindowInsets) support to Jetpack Compose.
 
+### 🍭 [System UI Controller](./systemuicontroller/)
+A library which provides easy-to-use utilities for updating the System UI bar colors within Jetpack Compose.
```
Comment (nickbutcher) [inline README.md:14]:
```suggestion
A library which provides easy-to-use utilities for updating the System UI (status and navigation bars) colors from Jetpack Compose.
```

File: docs/systemuicontroller.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
+# System UI Controller for Jetpack Compose
+
+[![Maven Central](https://img.shields.io/maven-central/v/com.google.accompanist/accompanist-systemuicontroller)](https://search.maven.org/search?q=g:com.google.accompanist)
+
+System UI Controller provides easy-to-use utilities for updating the System UI bar colors within Jetpack Compose.
+
+## Usage
+To setup Insets in your composables, you need to call the provide an instance to the `LocalSystemUiController`
```
Comment (nickbutcher) [inline docs/systemuicontroller.md:8]:
```suggestion
To control system UI in your composables, you must provide an instance to the `LocalSystemUiController`
```

File: docs/systemuicontroller.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+```
+
+Then in your layouts, you can update the system bar colors as necessary like so:
+
+``` kotlin
+@Composable
+fun MyHomeScreen() {
+    // Get the current SystemUiController
+    val systemUiController = LocalSystemUiController.current
+    val isSystemInDarkTheme = isSystemInDarkTheme()
```
Comment (nickbutcher) [inline docs/systemuicontroller.md:28]: What about querying the current theme? e.g.

```
val isDarkTheme = !MaterialTheme.colors.isLight
```

File: docs/systemuicontroller.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        )
+
+        // setStatusBarsColor() and setNavigationBarsColor() also exist
+    }
+}
+```
+
+## System bar icon colors
+The library automatically handles API level differences when running on Android devices. If we look at the example
+of status bar icons, Android only natively handles dark icons there on API 23+. This library handles this by automatically
```
Comment (nickbutcher) [inline docs/systemuicontroller.md:45]:
```suggestion
of status bar icons, Android only natively supports dark icons on API 23+. This library handles this by automatically
```

File: docs/systemuicontroller.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+altering the requested color with a scrim, to maintain contrast:
+
+![](api-scrim.png)
+
+Similar happens on navigation bar color, which is only available on API 26+.
+
+### Modifying scrim logic
+
+The scrim logic can be modified as needed by your needs:
```
Comment (nickbutcher) [inline docs/systemuicontroller.md:54]:
```suggestion
The scrim logic can be modified if needed:
```

File: docs/systemuicontroller.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+System UI Controller provides easy-to-use utilities for updating the System UI bar colors within Jetpack Compose.
+
+## Usage
+To setup Insets in your composables, you need to call the provide an instance to the `LocalSystemUiController`
+composition local. This would typically be done near the top level of your composable hierarchy:
+
+``` kotlin
+setContent {
+    // Create a controller, and provide it to the LocalSystemUiController
+    val controller = androidSystemUiController(LocalView.current)
```
Comment (nickbutcher) [inline docs/systemuicontroller.md:14]: This should be remembered right? Or should we offer a `rememberAndroidSBC` method?

File: sample/src/main/java/com/google/accompanist/sample/insets/EdgeToEdgeLazyColumn.kt [hunk 7]
```diff
@@ -69,73 +75,76 @@ class EdgeToEdgeLazyColumn : ComponentActivity() {
...
-@OptIn(ExperimentalStdlibApi::class)
 @Composable
 private fun Sample() {
-    ProvideWindowInsets {
-        Surface {
-            Box(Modifier.fillMaxSize()) {
-                // A state instance which allows us to track the size of the top app bar
-                var topAppBarSize by remember { mutableStateOf(0) }
+    val systemUiController = LocalSystemUiController.current
+    val isSystemInDarkTheme = isSystemInDarkTheme()
```
Comment (nickbutcher) [inline sample/src/main/java/com/google/accompanist/sample/insets/EdgeToEdgeLazyColumn.kt:92]: Again, query `MaterialTheme`?

File: systemuicontroller/src/androidTest/AndroidManifest.xml [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  ~
+  ~ Unless required by applicable law or agreed to in writing, software
+  ~ distributed under the License is distributed on an "AS IS" BASIS,
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
+    package="com.google.accompanist.insets.test">
```
Comment (nickbutcher) [inline systemuicontroller/src/androidTest/AndroidManifest.xml:18]:
```suggestion
    package="com.google.accompanist.systemuicontroller.test">
```

File: systemuicontroller/src/androidTest/AndroidManifest.xml [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+    <uses-sdk android:targetSdkVersion="30" />
+
+    <application>
+        <activity
+            android:name="com.google.accompanist.insets.InsetsTestActivity"
```
Comment (nickbutcher) [inline systemuicontroller/src/androidTest/AndroidManifest.xml:24]:
```suggestion
            android:name="com.google.accompanist.systemuicontroller.SystemUIControllerTestActivity"
```

File: systemuicontroller/src/androidTest/java/com/google/accompanist/systemuicontroller/SystemUiControllerTest.kt [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+            // We return red just to assert below.
+            Color.Red
+        }
+
+        // Assert that the colors were darkened color is not used
+        assertThat(Color(window.navigationBarColor)).isEqualTo(Color.White)
+    }
+}
+
+val AndroidComposeTestRule<*, *>.contentView
```
Comment (nickbutcher) [inline systemuicontroller/src/androidTest/java/com/google/accompanist/systemuicontroller/SystemUiControllerTest.kt:171]: Handy!

File: systemuicontroller/src/main/java/com/google/accompanist/systemuicontroller/SystemUiController.kt [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+     * API level that only supports white status bar icons.
+     * @param darkIcons Whether dark status bar icons would be preferable.
+     * @param transformColorForLightContent A lambda which will be invoked to transform [color] if
+     * dark icons were requested but are not available. Defaults to applying a black scrim.
+     */
+    fun setStatusBarColor(
+        color: Color,
+        darkIcons: Boolean = color.luminance() > 0.5f,
+        transformColorForLightContent: (Color) -> Color = BlackScrimmed
+    ) = Unit
```
Comment (nickbutcher) [inline systemuicontroller/src/main/java/com/google/accompanist/systemuicontroller/SystemUiController.kt:49]: Why `= Unit`?

File: systemuicontroller/src/main/java/com/google/accompanist/systemuicontroller/SystemUiController.kt [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+        color: Color,
+        darkIcons: Boolean,
+        transformColorForLightContent: (Color) -> Color
+    ) {
+        windowInsetsController?.isAppearanceLightStatusBars = darkIcons
+
+        window?.statusBarColor = when {
+            darkIcons && windowInsetsController?.isAppearanceLightStatusBars != true -> {
+                // If we're set to use dark icons, but our windowInsetsController call didn't
+                // succeed (usually to API level), we instead transform the color to maintain
```
Comment (nickbutcher) [inline systemuicontroller/src/main/java/com/google/accompanist/systemuicontroller/SystemUiController.kt:112]:
```suggestion
                // succeed (usually due to API level), we instead transform the color to maintain
```

File: systemuicontroller/src/main/java/com/google/accompanist/systemuicontroller/SystemUiController.kt [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+
+    val window = view.context.findWindow()
+    val windowInsetsController = ViewCompat.getWindowInsetsController(view)
+
+    override fun setStatusBarColor(
+        color: Color,
+        darkIcons: Boolean,
+        transformColorForLightContent: (Color) -> Color
+    ) {
+        windowInsetsController?.isAppearanceLightStatusBars = darkIcons
```
Comment (nickbutcher) [inline systemuicontroller/src/main/java/com/google/accompanist/systemuicontroller/SystemUiController.kt:107]: TIL!

File: systemuicontroller/src/main/java/com/google/accompanist/systemuicontroller/SystemUiController.kt [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+        color: Color,
+        darkIcons: Boolean,
+        transformColorForLightContent: (Color) -> Color
+    ) {
+        windowInsetsController?.isAppearanceLightNavigationBars = darkIcons
+
+        window?.navigationBarColor = when {
+            darkIcons && windowInsetsController?.isAppearanceLightNavigationBars != true -> {
+                // If we're set to use dark icons, but our windowInsetsController call didn't
+                // succeed (usually to API level), we instead transform the color to maintain
```
Comment (nickbutcher) [inline systemuicontroller/src/main/java/com/google/accompanist/systemuicontroller/SystemUiController.kt:130]:
```suggestion
                // succeed (usually due to API level), we instead transform the color to maintain
```

File: systemuicontroller/src/main/java/com/google/accompanist/systemuicontroller/SystemUiController.kt [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+    }
+}
+
+/**
+ * An [androidx.compose.runtime.CompositionLocalProvider] holding the current [LocalSystemUiController]. Defaults to a
+ * no-op controller; consumers should [provide][androidx.compose.runtime.CompositionLocalProvider] a real one.
+ */
+val LocalSystemUiController = staticCompositionLocalOf<SystemUiController> { NoOpSystemUiController }
+
+private val BlackScrim = Color(0f, 0f, 0f, 0.33f) // 20% opaque black
```
Comment (nickbutcher) [inline systemuicontroller/src/main/java/com/google/accompanist/systemuicontroller/SystemUiController.kt:154]: 33%?

**Review by nickbutcher — COMMENTED** [review]
Thanks for library-fying this!

Main comment is that I think that all creations of `androidSystemUiController` need to be remembered. Otherwise some small comments.

File: systemuicontroller/src/androidTest/java/com/google/accompanist/systemuicontroller/SystemUiControllerTest.kt [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+import androidx.test.annotation.UiThreadTest
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import androidx.test.filters.LargeTest
+import androidx.test.filters.SdkSuppress
+import com.google.common.truth.Truth.assertThat
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+
```
Comment (nickbutcher) [inline systemuicontroller/src/androidTest/java/com/google/accompanist/systemuicontroller/SystemUiControllerTest.kt:34]: Add tests for setting dark icons as well?

Comment (chrisbanes) [reply systemuicontroller/src/androidTest/java/com/google/accompanist/systemuicontroller/SystemUiControllerTest.kt:34]: Yep, the bottom 4 tests.

File: systemuicontroller/src/main/java/com/google/accompanist/systemuicontroller/SystemUiController.kt (hunk 13 above)
Comment (chrisbanes) [reply systemuicontroller/src/main/java/com/google/accompanist/systemuicontroller/SystemUiController.kt:107]: New in AndroidX Core 1.5.0-beta02

File: systemuicontroller/src/main/java/com/google/accompanist/systemuicontroller/SystemUiController.kt (hunk 15 above)
Comment (chrisbanes) [reply systemuicontroller/src/main/java/com/google/accompanist/systemuicontroller/SystemUiController.kt]: fixed comment

File: docs/systemuicontroller.md (hunk 3 above)
Comment (chrisbanes) [reply docs/systemuicontroller.md:28]: Done

File: systemuicontroller/build.gradle [hunk 17]
```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2020 The Android Open Source Project
```
Comment (nickbutcher) [inline systemuicontroller/build.gradle:2]:
```suggestion
 * Copyright 2021 The Android Open Source Project
```

File: systemuicontroller/src/androidTest/AndroidManifest.xml [hunk 18]
```diff
@@ -0,0 +1,28 @@
+<!--
+  ~ Copyright 2020 The Android Open Source Project
```
Comment (nickbutcher) [inline systemuicontroller/src/androidTest/AndroidManifest.xml:2]:
```suggestion
  ~ Copyright 2021 The Android Open Source Project
```

File: systemuicontroller/src/androidTest/java/com/google/accompanist/systemuicontroller/SystemUiControllerTest.kt [hunk 19]
```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2020 The Android Open Source Project
```
Comment (nickbutcher) [inline systemuicontroller/src/androidTest/java/com/google/accompanist/systemuicontroller/SystemUiControllerTest.kt:2]:
```suggestion
 * Copyright 2021 The Android Open Source Project
```

File: systemuicontroller/src/main/AndroidManifest.xml [hunk 20]
```diff
@@ -0,0 +1,18 @@
+<!--
+  ~ Copyright 2020 The Android Open Source Project
```
Comment (nickbutcher) [inline systemuicontroller/src/main/AndroidManifest.xml:2]:
```suggestion
  ~ Copyright 2021 The Android Open Source Project
```

File: systemuicontroller/src/androidTest/java/com/google/accompanist/systemuicontroller/SystemUiControllerTest.kt (hunk 16 above)
Comment (chrisbanes) [reply systemuicontroller/src/androidTest/java/com/google/accompanist/systemuicontroller/SystemUiControllerTest.kt:35]: Ah, done

File: README.md [hunk 21]
```diff
@@ -10,6 +10,9 @@ A number of libraries which integrate popular image loading libraries into Jetpa
 ### 📐 [Insets](./insets/)
 A library which brings [WindowInsets](https://developer.android.com/reference/kotlin/android/view/WindowInsets) support to Jetpack Compose.
 
+### 🍭 [System UI Controller](./systemuicontroller/)
```
Comment (nickbutcher) [inline README.md:13]:
```suggestion
### 🍫 [System UI Controller](./systemuicontroller/)
```

Geddit? Cus its about `bar`s!

Comment (chrisbanes) [reply README.md]: Surreeeeee 😆

File: systemuicontroller/src/androidTest/java/com/google/accompanist/systemuicontroller/SystemUiControllerTest.kt (hunk 16 above)
Comment (nickbutcher) [reply systemuicontroller/src/androidTest/java/com/google/accompanist/systemuicontroller/SystemUiControllerTest.kt:35]: Thanks!

**Pull request merged**